decode.go

```diff
@@ -1425,7 +1425,13 @@ func (d *Decoder) decodeStruct(ctx context.Context, dst reflect.Value, src ast.N
 	}
 
 	if d.validator != nil {
-		if err := d.validator.Struct(dst.Interface()); err != nil {
+		var err error
```

Comment:
As you already mentioned in your comment, `dst` is guaranteed to be addressable here, so there's no need to call `CanAddr`.

```
if err := d.validator.Struct(dst.Addr().Interface()); err != nil {
```